power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -598,17 +598,130 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
     static constexpr auto transformer_index_of = container_impl::get_cls_pos_v<T, TransformerTypes...>;
     static_assert(((transformer_index_of<TransformerTypes> < sizeof...(TransformerTypes)) && ...));
 
+    struct BinarySearch {
+      public:
+        BinarySearch() = default;
+        BinarySearch(IntS tap_pos, IntS tap_min, IntS tap_max) { reset_bs(tap_pos, tap_min, tap_max); }
+
+        IntS get_bs_current_tap() const { return current; }
+        bool get_bs_last_down() const { return last_down; }
+        bool get_bs_inevitable_run() const { return inevitable_run; }
+        bool get_end_of_bs() const { return get_bs_tap_lower_bound() >= get_bs_tap_upper_bound(); }
+
+        void set_bs_current_tap(IntS _current_tap) { this->current = _current_tap; }
+        void set_bs_last_check(bool _last_check) { this->last_check = _last_check; }
+        void set_bs_inevitable_run(bool _inevitable_run) { this->inevitable_run = _inevitable_run; }
+
+        void recalibrate(bool strategy_max) {
+            if (get_bs_tap_reverse() ? strategy_max : !strategy_max) {
+                set_bs_tap_lower_bound(get_bs_current_tap());
+                set_bs_last_down(false);
+            } else {
+                set_bs_tap_upper_bound(get_bs_current_tap());
+                set_bs_last_down(true);
+            }
+        }
+
+        void propose_new_pos(bool strategy_max, bool above_range) {
+            auto is_down = above_range ? get_bs_tap_reverse() : !get_bs_tap_reverse();
+            if (get_bs_last_check()) {
+                set_bs_current_tap(is_down ? get_bs_tap_lower_bound() : get_bs_tap_upper_bound());
+                set_bs_inevitable_run(true);
+            } else {
+                set_bs_last_down(is_down);
+                adjust_bs(strategy_max);
+            }
+        }
+
+        IntS post_process(bool strategy_max, bool previous_down, bool& tap_changed) {
+            const bool prefer_higher = strategy_max ? !get_bs_tap_reverse() : get_bs_tap_reverse();
+            const auto tap_pos = search_bs(prefer_higher);
+            const auto tap_diff = tap_pos - get_bs_current_tap();
+            if (tap_diff == 0) {
+                tap_changed = false;
+                return tap_pos;
+            }
+            if ((tap_diff == 1 && previous_down) || (tap_diff == -1 && !previous_down)) {
+                set_bs_last_check(true);
+            }
+            tap_changed = true;
+            set_bs_current_tap(tap_pos);
+            return tap_pos;
+        }
+
+      private:
+        IntS lower_bound{-12};      // tap position lower bound
+        IntS upper_bound{12};       // tap position upper bound
```

Comment:
why specifically `12`? sounds like this could become a bug. i would propose to use
```suggestion
        IntS lower_bound{std::numeric_limits<IntS>::lowest()};      // tap position lower bound
        IntS upper_bound{std::numeric_limits<IntS>::max()};       // tap position upper bound
```

Reply:
I'm fine with either. It's a private structure that's only used in one way throughout the code, so they are always initialized with meaningful input.

Reply:
comment addressed in the latest commit.